pkg/server/message.go

```diff
@@ -254,69 +280,154 @@ func (t *TranslateServiceServer) UpdateMessages(
 
 // helpers
 
-// populateTranslatedMessages adds any missing messages from the original messages to all translated messages.
-func (t *TranslateServiceServer) populateTranslatedMessages(
-	ctx context.Context,
-	serviceID uuid.UUID,
-	originalMessages *model.Messages,
-	allMessages []model.Messages,
-) error {
+// alterTranslations alter the status of specific messages in multiple languages to "untranslated"
+// based on a list of message IDs, while keeping the "original" language messages unchanged.
+func (t *TranslateServiceServer) alterTranslations(
+	all model.MessagesSlice,
+	untranslatedIDs []string,
+) model.MessagesSlice {
+	all = all.Clone()
+
+	if len(untranslatedIDs) == 0 {
+		return all
+	}
+
+	// Return if only one language is present
+	if len(all) == 1 {
+		return all
+	}
+
+	original, others := all.SplitOriginal()
+
+	// Return if messages don't contain original language
+	if original == nil {
+		return all
+	}
+
+	// Update altered messages for all translations
+	for _, msg := range others {
+		for i := range msg.Messages {
+			if slices.Contains(untranslatedIDs, msg.Messages[i].ID) {
```

Comment:
outside `for` we can sort `untranslatedIDs`, afterwards we can use binary search to determine if the slice contains the value

```go
slices.Sort(untranslatedIDs)
```

```suggestion
			if _, found := slices.BinarySearch(untranslatedIDs, msg.Messages[i].ID); found {
```